--- a/usr/share/lintian/frontend/dplint	2019-08-10 17:05:00.389258614 +0200
+++ b/usr/share/lintian/frontend/dplint	2019-08-10 17:05:08.973549548 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env -S perl -w
 #
 # dplint -- General purpose frontend for Debian package checker
 #
--- a/usr/share/lintian/helpers/coll/file-info-helper	2019-08-10 17:05:13.385699081 +0200
+++ b/usr/share/lintian/helpers/coll/file-info-helper	2019-08-10 17:05:19.053891181 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 
 no lib '.';
 
--- a/usr/share/lintian/helpers/coll/objdump-info-helper	2019-08-10 17:05:30.762287986 +0200
+++ b/usr/share/lintian/helpers/coll/objdump-info-helper	2019-08-10 17:05:36.114469373 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl
+#!/usr/bin/env perl
 # objdump-info-helper -- lintian collection script
 
 # Most of it is taken from objdump-info (Lintian 2.5.9), which had the
--- a/usr/share/lintian/helpers/coll/strings-helper	2019-08-10 17:05:22.217998415 +0200
+++ b/usr/share/lintian/helpers/coll/strings-helper	2019-08-10 17:05:27.422174788 +0200
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -w
+#!/usr/bin/env -S perl -w
 # strings -- lintian collection script
 
 # Copyright (C) 2012 Niels Thykier <niels@thykier.net>
