From 72651a6e0f25eac5df937e539b94c3d9221cba38 Mon Sep 17 00:00:00 2001
From: Mathieu Schroeter <mathieu@schroetersa.ch>
Date: Wed, 13 Oct 2021 14:22:20 +0200
Subject: Extra msys2 patches


diff --git a/Lib/cgi.py b/Lib/cgi.py
index 5b903e0..1237de2 100755
--- a/Lib/cgi.py
+++ b/Lib/cgi.py
@@ -1,4 +1,4 @@
-#! /usr/local/bin/python
+#!/usr/bin/env python2
 
 # NOTE: the above "/usr/local/bin/python" is NOT a mistake.  It is
 # intentionally NOT "/usr/bin/env python".  On many systems
diff --git a/Lib/distutils/tests/test_build_scripts.py b/Lib/distutils/tests/test_build_scripts.py
index ad3c1a2..1562a62 100644
--- a/Lib/distutils/tests/test_build_scripts.py
+++ b/Lib/distutils/tests/test_build_scripts.py
@@ -55,7 +55,7 @@ class BuildScriptsTestCase(support.TempdirManager,
         expected = []
         expected.append("script1.py")
         self.write_script(dir, "script1.py",
-                          ("#! /usr/bin/env python2.3\n"
+                          ("#! /usr/bin/env python2\n"
                            "# bogus script w/ Python sh-bang\n"
                            "pass\n"))
         expected.append("script2.py")
diff --git a/Lib/distutils/tests/test_install_scripts.py b/Lib/distutils/tests/test_install_scripts.py
index 9c0ba6d..c81e36e 100644
--- a/Lib/distutils/tests/test_install_scripts.py
+++ b/Lib/distutils/tests/test_install_scripts.py
@@ -48,7 +48,7 @@ class InstallScriptsTestCase(support.TempdirManager,
             finally:
                 f.close()
 
-        write_script("script1.py", ("#! /usr/bin/env python2.3\n"
+        write_script("script1.py", ("#! /usr/bin/env python2\n"
                                     "# bogus script w/ Python sh-bang\n"
                                     "pass\n"))
         write_script("script2.py", ("#!/usr/bin/python\n"
diff --git a/Parser/pgen.stamp b/Parser/pgen.stamp
new file mode 100644
index 0000000..e16c76d
--- /dev/null
+++ b/Parser/pgen.stamp
@@ -0,0 +1 @@
+""
diff --git a/setup.py b/setup.py
index 3ceea1c..6880ae9 100644
--- a/setup.py
+++ b/setup.py
@@ -1333,7 +1333,6 @@ class PyBuildExt(build_ext):
             sqlite_defines = [('MODULE_NAME', '"sqlite3"')]
 
             # Comment this out if you want the sqlite3 module to be able to load extensions.
-            sqlite_defines.append(("SQLITE_OMIT_LOAD_EXTENSION", "1"))
 
             if host_platform == 'darwin':
                 # In every directory on the search path search for a dynamic
