--- a/wpkg/libdebpackages/wpkg_control_fields.cpp	Fri Aug  8 15:39:36 2014
+++ b/wpkg/libdebpackages/wpkg_control_fields.cpp	Fri Aug  8 15:41:50 2014
@@ -881,7 +881,7 @@
                 .arg(get_filename())
                 .arg(get_line())
                 .arg(get_name())
-                .arg(len)
+                .arg(static_cast<int>(len))
             .level(wpkg_output::level_error)
             .module(wpkg_output::module_control)
             .package(f_field_file->get_package_name())
