// shell-craft/index.js
'use strict';

var async    = require ('async');


function Prompt () {
  return [{
    type    : 'input',
    name    : 'command',
    message : '$'
  }];
}

function Command (handler) {
  this._params  = {};
  this._desc    = null;
  this._handler = handler;
  return this;
}

Command.prototype.call = function () {
  return this._handler.apply (this, arguments);
};

function Commands () {
}

Commands.prototype._add = function (cmd, obj) {
  Commands.prototype[cmd] = obj;
};

function ShellCraft () {
  var self = this;

  this._exit = false;

  this.commands = new Commands ();
  this.commands._add ('exit', new Command (function () {
    self._exit = true;
  }));
  this.commands._add ('help', new Command (function () {
    for (var fct in self.commands) {
      if (!self.commands.hasOwnProperty (fct) && !/^_/.test(fct)) {
        console.log (fct);
      }
    }
  }));

  this.options = {};
  this.prompt  = new Prompt ();
}

ShellCraft.prototype.isExit = function () {
  return this._exit;
};

/**
 * Register external commands.
 */
ShellCraft.prototype.registerExtension = function (shellExt) {
  var extension = require (shellExt) ();

  extension.forEach (function (cmd) {
    this.commands._add (cmd, extension[cmd]);
  });
};

/**
 * @api private
 */
ShellCraft.prototype.shell = function (callback) {
  var self = this;

  var inquirer = require ('inquirer');

  async.forever (function (next) {
    inquirer.prompt (self.prompt, function (answers) {
      var cmdArgs = answers.command.split (' ');
      var cmd = cmdArgs[0];
      cmdArgs.shift ();

      try {
        self.commands[cmd].call (cmdArgs);
      } catch (ex) {
        if (answers.command.length) {
          console.log ('command ' + cmd + ' unknown');
        }
      }
      next (self.isExit () ? 'good bye' : null);
    });
  }, function (results) {
    if (callback) {
      callback (results);
    }
  });
};

/**
 * @api private
 */
ShellCraft.prototype.cli = function (callback) {
  var program = require ('commander');

  program.version ('0.1.0');

  // TODO: work in progress..
};

ShellCraft.prototype.begin = function (callback) {
  /* Run the Shell. */
  if (process.argv.length === 2) {
    this.shell (callback);
    return;
  }

  /* Run in command line. */
  this.cli (callback);
};

exports = module.exports = new ShellCraft ();
