--- a/networking/tls_aesgcm.c        2023-01-05 16:29:10.000000000 +0100
+++ b/networking/tls_aesgcm.c       2025-03-13 10:30:25.971495891 +0100
@@ -6,6 +6,30 @@
 
 #include "tls.h"
 
+/* Number of bits in a word of type `long int'.  */
+#ifdef LONG_MAX
+# if LONG_MAX == 2147483647
+#  define LONG_BIT     32
+# else
+/* Safe assumption.  */
+#  define LONG_BIT     64
+# endif
+#elif defined __LONG_MAX__
+# if __LONG_MAX__ == 2147483647
+#  define LONG_BIT     32
+# else
+/* Safe assumption.  */
+#  define LONG_BIT     64
+# endif
+#else
+# include <bits/wordsize.h>
+# if __WORDSIZE == 64
+#  define LONG_BIT     64
+# else
+#  define LONG_BIT     32
+# endif
+#endif
+
 typedef uint8_t byte;
 typedef uint32_t word32;
 #define XMEMSET memset
